File: Voci/Sources/Reminders/ReminderContextGate.swift
// Sources/Reminders/ReminderContextGate.swift — voice-suppression gate (contract §B, T072).
// Constitution I / the contract's explicit instruction: "fail toward SUPPRESSED when unsure."
// Every concrete signal below either (a) produces a definite true/false from a public,
// no-extra-entitlement API, or (b) is an injectable extension point that defaults to "no signal"
// (not suppressing on that axis alone) until something wires it — see the doc comment on each
// property for exactly which is which, and this task's final report for what's left unwired.
import Foundation
import AVFoundation

@MainActor
final class ReminderContextGate {
    /// Calendar-busy windows. Injected by the caller — `[]` until Phase 3's calendar integration
    /// lands (contract §B says so explicitly: "busyIntervals injected — [] until P3"). A `now`
    /// inside any interval here suppresses voice.
    var busyIntervals: [DateInterval] = []

    // MARK: - Extension points
    //
    // These four cover signals this gate has no live-instance access to from its own files (mic
    // capture belongs to `Sources/Speech/SpeechCapture.swift`, owned elsewhere; screen-sharing and
    // Focus/DND have no stable, unprivileged public API on macOS as of this SDK — Focus status
    // specifically requires the restricted `com.apple.developer.focus-status` entitlement via
    // `INFocusStatusCenter`, which this task deliberately does NOT add speculatively to
    // `Resources/Voci.entitlements`, since that entitlement needs Apple approval and this is a
    // MAS-first app — research.md R3). `nil` means "not wired yet" and reads as "no signal" for
    // that one axis; once wired (by whichever agent owns the live instance/API), a `true` from any
    // of these suppresses voice same as a concrete check would.
    var isLocalMicCaptureActive: (() -> Bool)?
    var isOtherAudioPlaying: (() -> Bool)?
    var isScreenBeingShared: (() -> Bool)?
    var isDoNotDisturbOn: (() -> Bool)?

    init() {}

    /// `true` suppresses the voice rung for this fire (visual delivery is unaffected either way —
    /// this gate only ever guards `VoiceReminderChannel`, never `UNUserNotificationCenter`).
    func shouldSuppressVoice(now: Date) -> Bool {
        if busyIntervals.contains(where: { $0.contains(now) }) { return true }
        if isLocalMicCaptureActive?() == true { return true }
        if isOtherAudioPlaying?() == true { return true }
        if isScreenBeingShared?() == true { return true }
        if isDoNotDisturbOn?() == true { return true }
        if isAnotherAppUsingMicrophone() { return true }
        return false
    }

    /// Best-effort "a call is likely active" signal with NO extra entitlement and NO permission
    /// prompt of its own: `AVCaptureDevice.isInUseByAnotherApplication` is a standard AVFoundation
    /// capture-device property that apps use to detect camera/mic contention (e.g. before grabbing
    /// the mic themselves) — reading it does not itself request microphone access. Fails toward
    /// suppressed: any state other than a definite "no device / not in use" suppresses.
    private func isAnotherAppUsingMicrophone() -> Bool {
        guard let device = AVCaptureDevice.default(for: .audio) else {
            // No audio input device at all — nothing else could plausibly be "on a call" via it.
            return false
        }
        return device.isInUseByAnotherApplication
    }
}

File: windows/src/Volar.Reminders/ReminderContextGate.cs
// ReminderContextGate.cs — port of Sources/Reminders/ReminderContextGate.swift: voice-suppression
// gate (contract §B). Constitution I / the contract's explicit instruction: "fail toward
// SUPPRESSED when unsure." Every concrete signal below either (a) produces a definite true/false
// from an injected extension point, or (b) defaults to "no signal" (not suppressing on that axis
// alone) until a caller wires it — see the doc comment on each property.
namespace Volar.Reminders;

/// <summary>A closed time window, mirroring Swift's <c>DateInterval</c> (used here for busy-calendar
/// windows).</summary>
public readonly record struct ReminderTimeRange(DateTimeOffset Start, DateTimeOffset End)
{
    public bool Contains(DateTimeOffset instant) => instant >= Start && instant <= End;
}

public sealed class ReminderContextGate
{
    /// <summary>
    /// Calendar-busy windows. Injected by the caller — empty until calendar integration lands
    /// (contract §B says so explicitly: "busyIntervals injected — [] until P3"). A `now` inside any
    /// interval here suppresses voice.
    /// </summary>
    public IReadOnlyList<ReminderTimeRange> BusyIntervals { get; set; } = Array.Empty<ReminderTimeRange>();

    // MARK: - Extension points
    //
    // The first four mirror the Swift original 1:1 (mic capture / other audio / screen-sharing /
    // Focus-DND all have no stable, unprivileged public API this pure project can call directly —
    // same reasoning Swift gave for leaving them as `nil`-by-default injection points).
    public Func<bool>? IsLocalMicCaptureActive { get; set; }

    public Func<bool>? IsOtherAudioPlaying { get; set; }

    public Func<bool>? IsScreenBeingShared { get; set; }

    public Func<bool>? IsDoNotDisturbOn { get; set; }

    /// <summary>
    /// Windows deviation (documented, not a gap): Swift's fifth signal
    /// (<c>isAnotherAppUsingMicrophone</c>) was a CONCRETE, no-extra-entitlement AVFoundation check
    /// (<c>AVCaptureDevice.isInUseByAnotherApplication</c>) called directly from this file. There is
    /// no framework-free Windows equivalent this pure project can call without violating the "no
    /// Windows App SDK" dependency rule, so it is demoted here from "concrete check" to a fifth
    /// injectable extension point — defaults to no signal (not suppressing), same as the other
    /// four, until a Wave-3 caller wires a real WASAPI/MediaCapture-backed check.
    /// </summary>
    public Func<bool>? IsAnotherAppUsingMicrophone { get; set; }

    /// <summary>
    /// <see langword="true"/> suppresses the voice rung for this fire (visual delivery is
    /// unaffected either way — this gate only ever guards the spoken channel, never the toast
    /// banner).
    /// </summary>
    public bool ShouldSuppressVoice(DateTimeOffset now)
    {
        foreach (var range in BusyIntervals)
        {
            if (range.Contains(now))
            {
                return true;
            }
        }
        if (IsLocalMicCaptureActive?.Invoke() == true)
        {
            return true;
        }
        if (IsOtherAudioPlaying?.Invoke() == true)
        {
            return true;
        }
        if (IsScreenBeingShared?.Invoke() == true)
        {
            return true;
        }
        if (IsDoNotDisturbOn?.Invoke() == true)
        {
            return true;
        }
        if (IsAnotherAppUsingMicrophone?.Invoke() == true)
        {
            return true;
        }
        return false;
    }
}
